File: Programmers/피로도.cpp
// level 2
#include <string>
#include <vector>

using namespace std;

int visited[9] = {0, };
vector<vector<int>> per;

int max(int a, int b) {
    return a > b ? a : b; 
}

int cal_cnt(int size, int k) {
    int cnt = 0, cur = k;
    
    for (auto p : per) {
        if (p[0] <= cur && cur > 0) {
            cur -= p[1];
            cnt++;
        }
        
        else {
            break;
        }
    }
    
    return cnt;
}

void DFS(int size, int k, vector<vector<int>> &dungeons, int level, int &answer) {
    if (level == size) {
        int ret = cal_cnt(size, k);
        answer = max(answer, ret);
        return;
    }
    
    for (int i = 0; i < size; ++i) {
        if (visited[i] == 1)
            continue;
        visited[i] = 1;
        per.push_back(dungeons[i]);
        DFS(size, k, dungeons, level + 1, answer);
        per.pop_back();
        visited[i] = 0;
    }
}

int solution(int k, vector<vector<int>> dungeons) {
    int answer = -1;
    int size = dungeons.size();
    
    DFS(size, k, dungeons, 0, answer);
    return answer;
}

File: Programmers/피로도.py
# level 2
import itertools

def solution(k, dungeons):
    answer = -1
    # 순열을 이용
    per = itertools.permutations(dungeons, len(dungeons))
    
    # 완전탐색
    # 한 순열 하나씩 확인
    for candi_p in per:
        cur_p = k # 초기화
        cnt = 0
        for c_p in candi_p:
            # 던전을 지날 수 있으면
            if c_p[0] <= cur_p:
                cnt += 1
                cur_p -= c_p[1]
        answer = max(answer, cnt)
    return answer